WordPress/Classes/ViewRelated/Themes/ThemeBrowserViewController.swift

```diff
@@ -0,0 +1,220 @@
+        fetchRequest.sortDescriptors = [sort]
+        let frc = NSFetchedResultsController(fetchRequest: fetchRequest,
+            managedObjectContext: self.managedObjectContext,
+            sectionNameKeyPath: nil,
+            cacheName: nil)
+        frc.delegate = self
+        
+        do {
+            try frc.performFetch()
+        } catch let error as NSError {
+            DDLogSwift.logError("Error fetching themes: \(error.localizedDescription)")
+        }
+        
+        return frc
+    }()
+    
+    /**
+     *  @brief      The themes service we'll use in this VC and its helpers
+     */
+    private lazy var themeService : ThemeService = {
+        ThemeService(managedObjectContext: self.managedObjectContext)
+    }()
+    private lazy var managedObjectContext = {
+        ContextManager.sharedInstance().mainContext
+    }()
+    private var retiredThemes = Set<Theme>()
+    private var fetchAnimation = false
+   
+    // MARK: - UIViewController
+
+    public override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        title = NSLocalizedString("Themes", comment: "Title of Themes browser page")
+        
+        WPStyleGuide.configureColorsForView(view, collectionView:collectionView)
+        
+        updateThemes()
+    }
+
+    public override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
+        super.viewWillTransitionToSize(size, withTransitionCoordinator: coordinator)
+        
+        collectionView?.collectionViewLayout.invalidateLayout()
+    }
+
+    // MARK: - Updating the list of themes
+    
+    private func updateThemes() {
+        if let fetchedThemes = themesController.fetchedObjects as? [Theme] where !fetchedThemes.isEmpty {
+            retiredThemes = retiredThemes.union(Set(fetchedThemes))
+        } else {
+            fetchAnimation = true
+            let title = NSLocalizedString("Fetching Themes...", comment:"Text displayed while fetching themes")
+            WPNoResultsView.displayAnimatedBoxWithTitle(title, message: nil, view: self.view)
+        }
+        
+        updateThemePage(1)
+    }
+
+    private func updateThemePage(page: NSInteger) {
+        assert(page > 0)
+        
+        themeService.getThemesForBlog(blog,
+            page: page,
+            success: { [weak self] (themes: [Theme]?, hasMore: Bool) in
+                guard let strongSelf = self else {
+                    return
+                }
+                
+                if let updatedThemes = themes where !updatedThemes.isEmpty {
+                    strongSelf.retiredThemes = strongSelf.retiredThemes.subtract(updatedThemes)
+                }
+                
+                if (hasMore) {
+                    strongSelf.updateThemePage(page + 1)
+                } else if !strongSelf.retiredThemes.isEmpty {
+                    let retireContext = strongSelf.managedObjectContext
```

Comment:
> I think the reader also does that, but it may belong to the service layer?

Yep. The reader does trigger a bit of clean up in `ReaderViewController cleanupPreviewedPostsAndTopics`. The actual work is handled by the service layer.